From 52227243949f372edd9efd82ad03250e67bdcc0d Mon Sep 17 00:00:00 2001
From: jiangheng <jiangheng14@huawei.com>
Date: Thu, 11 Jan 2024 16:19:06 +0800
Subject: [PATCH] 0006-dpdk-fix-dpdk-coredump-problem.patch

---
 lib/eal/linux/eal_interrupts.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/eal/linux/eal_interrupts.c b/lib/eal/linux/eal_interrupts.c
index d4919df..e681628 100644
--- a/lib/eal/linux/eal_interrupts.c
+++ b/lib/eal/linux/eal_interrupts.c
@@ -1129,7 +1129,7 @@ eal_intr_thread_main(__rte_unused void *arg)
 		 */
 		if (epoll_ctl(pfd, EPOLL_CTL_ADD, intr_pipe.readfd,
 						&pipe_event) < 0) {
-			rte_panic("Error adding fd to %d epoll_ctl, %s\n",
+			RTE_LOG(ERR, EAL, "Error adding fd to %d epoll_ctl, %s\n",
 					intr_pipe.readfd, strerror(errno));
 		}
 		numfds++;
@@ -1151,7 +1151,7 @@ eal_intr_thread_main(__rte_unused void *arg)
 			 */
 			if (epoll_ctl(pfd, EPOLL_CTL_ADD,
 					rte_intr_fd_get(src->intr_handle), &ev) < 0) {
-				rte_panic("Error adding fd %d epoll_ctl, %s\n",
+				RTE_LOG(ERR, EAL, "Error adding fd %d epoll_ctl, %s\n",
 					rte_intr_fd_get(src->intr_handle),
 					strerror(errno));
 			}
-- 
2.33.0

