From a76d953f02beecc057c96159e32e292b847b2d63 Mon Sep 17 00:00:00 2001
From: Changsheng Wu <wuchangsheng2@huawei.com>
Date: Fri, 12 Nov 2021 17:18:23 +0800
Subject: [PATCH] huawei-0006-dpdk-fix-dpdk-coredump-problem

---
 lib/eal/linux/eal_interrupts.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/eal/linux/eal_interrupts.c b/lib/eal/linux/eal_interrupts.c
index 6e3925efd4..621e43626e 100644
--- a/lib/eal/linux/eal_interrupts.c
+++ b/lib/eal/linux/eal_interrupts.c
@@ -1137,7 +1137,7 @@ eal_intr_thread_main(__rte_unused void *arg)
 		 */
 		if (epoll_ctl(pfd, EPOLL_CTL_ADD, intr_pipe.readfd,
 						&pipe_event) < 0) {
-			rte_panic("Error adding fd to %d epoll_ctl, %s\n",
+			RTE_LOG(ERR, EAL, "Error adding fd to %d epoll_ctl, %s\n",
 					intr_pipe.readfd, strerror(errno));
 		}
 		numfds++;
@@ -1159,7 +1159,7 @@ eal_intr_thread_main(__rte_unused void *arg)
 			 */
 			if (epoll_ctl(pfd, EPOLL_CTL_ADD,
 					rte_intr_fd_get(src->intr_handle), &ev) < 0) {
-				rte_panic("Error adding fd %d epoll_ctl, %s\n",
+				RTE_LOG(ERR, EAL, "Error adding fd %d epoll_ctl, %s\n",
 					rte_intr_fd_get(src->intr_handle),
 					strerror(errno));
 			}
-- 
2.27.0

