From 9cb1787c85a297b2400f820487a5a0a65ca34005 Mon Sep 17 00:00:00 2001
From: liya51 <luckly0822@qq.com>
Date: Fri, 11 Jul 2025 14:42:31 +0800
Subject: [PATCH] fix the secondary process core dumped project

---
 drivers/net/hinic3/hinic3_pmd_ethdev.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/net/hinic3/hinic3_pmd_ethdev.c b/drivers/net/hinic3/hinic3_pmd_ethdev.c
index ece8252..037e9ea 100644
--- a/drivers/net/hinic3/hinic3_pmd_ethdev.c
+++ b/drivers/net/hinic3/hinic3_pmd_ethdev.c
@@ -1894,6 +1894,17 @@ static void hinic3_dev_close(struct rte_eth_dev *eth_dev)
 {
 	struct hinic3_nic_dev *nic_dev =
 		HINIC3_RTE_ETH_DEV_TO_PRIVATE_NIC_DEV(eth_dev);
+	
+	if (rte_eal_process_type() != RTE_PROC_PRIMARY) {
+		PMD_DRV_LOG(INFO, "close %s in secondary process",
+			    eth_dev->data->name);
+#ifdef DPDK_20_11
+		return 0;
+#else
+		return;
+#endif
+	}
+
 #ifdef DPDK_20_11
 	int ret;
 #endif
-- 
2.45.1.windows.1

