diff --git a/drivers/net/hinic3/hinic3_pmd_fdir.c b/drivers/net/hinic3/hinic3_pmd_fdir.c
index 59721e0..50424f4 100644
--- a/drivers/net/hinic3/hinic3_pmd_fdir.c
+++ b/drivers/net/hinic3/hinic3_pmd_fdir.c
@@ -412,6 +412,7 @@ static void hinic3_fdir_tcam_ipv6_vxlan_init(struct rte_eth_dev *dev,
 			struct hinic3_tcam_key *tcam_key)
 {
 	struct hinic3_nic_dev *nic_dev = HINIC3_ETH_DEV_TO_PRIVATE_NIC_DEV(dev);
+	u8 bifur_en;
 
 	tcam_key->key_mask_ipv6.ip_proto = rule->key_mask.proto;
 	tcam_key->key_info_ipv6.ip_proto = rule->key_spec.proto;
@@ -423,8 +424,15 @@ static void hinic3_fdir_tcam_ipv6_vxlan_init(struct rte_eth_dev *dev,
 	tcam_key->key_info_ipv6.outer_ip_type = HINIC3_FDIR_IP_TYPE_IPV6;
 
 	tcam_key->key_mask_ipv6.function_id = HINIC3_UINT15_MAX;
-	tcam_key->key_info_ipv6.function_id =
-		hinic3_global_func_id(nic_dev->hwdev) & HINIC3_UINT15_MAX;
+	if (hinic3_get_bifur_enable(nic_dev->hwdev, &bifur_en, 0, 0) != 0) {
+		PMD_DRV_LOG(ERR, "hinic3 get port table bifur enable status failed.");
+	}
+	if (bifur_en)
+		tcam_key->key_info_ipv6.function_id = HINIC3_UINT15_MAX;
+	else {
+		tcam_key->key_info_ipv6.function_id =
+			hinic3_global_func_id(nic_dev->hwdev) & HINIC3_UINT15_MAX;
+	}
 
 	tcam_key->key_mask_ipv6.dport = rule->key_mask.dst_port;
 	tcam_key->key_info_ipv6.dport = rule->key_spec.dst_port;
@@ -441,6 +449,8 @@ static void hinic3_fdir_tcam_vxlan_init(struct rte_eth_dev *dev,
 			struct hinic3_tcam_key *tcam_key)
 {
 	struct hinic3_nic_dev *nic_dev = HINIC3_ETH_DEV_TO_PRIVATE_NIC_DEV(dev);
+	u8 bifur_en, iso_en;
+	u8 er_id = nic_dev->hwdev->cfg_mgmt->svc_cap.er_id;
 
 	if (rule->outer_ip_type == HINIC3_FDIR_IP_TYPE_IPV6) {
 		hinic3_fdir_tcam_ipv6_vxlan_init(dev, rule, tcam_key);
@@ -487,8 +497,19 @@ static void hinic3_fdir_tcam_vxlan_init(struct rte_eth_dev *dev,
 	tcam_key->key_info.tunnel_type = HINIC3_FDIR_TUNNEL_MODE_VXLAN;
 
 	tcam_key->key_mask.function_id = HINIC3_UINT15_MAX;
-	tcam_key->key_info.function_id =
-		hinic3_global_func_id(nic_dev->hwdev) & HINIC3_UINT15_MAX;
+	if (hinic3_get_bifur_enable(nic_dev->hwdev, &bifur_en, &iso_en, 0) != 0) {
+		PMD_DRV_LOG(ERR, "hinic3 get port table bifur enable status failed.");
+	}
+	if (bifur_en)
+		tcam_key->key_info.function_id = HINIC3_UINT15_MAX;
+	else {
+		tcam_key->key_info.function_id =
+			hinic3_global_func_id(nic_dev->hwdev) & HINIC3_UINT15_MAX;
+	}
+	if (iso_en) {
+		tcam_key->key_info.er_id = er_id;
+		tcam_key->key_info.er_id = HINIC3_UINT4_MAX;
+	}
 
 	if (rule->ip_type == HINIC3_FDIR_IP_TYPE_IPV4)
 		hinic3_fdir_tcam_vxlan_ipv4_init(rule, tcam_key);
