From ea382787ef02ca87a78555b41d0d0e0e0d8e24a7 Mon Sep 17 00:00:00 2001
From: thy <19855679967@163.com>
Date: Thu, 24 Apr 2025 19:18:34 +0800
Subject: [PATCH] fix support different queue depth

---
 drivers/net/hinic3/hinic3_pmd_ethdev.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/net/hinic3/hinic3_pmd_ethdev.c b/drivers/net/hinic3/hinic3_pmd_ethdev.c
index 0fceeb41bc..7d3442a951 100644
--- a/drivers/net/hinic3/hinic3_pmd_ethdev.c
+++ b/drivers/net/hinic3/hinic3_pmd_ethdev.c
@@ -640,6 +640,12 @@ static int hinic3_rx_queue_setup(struct rte_eth_dev *dev, uint16_t qid,
 
 	nic_dev = HINIC3_ETH_DEV_TO_PRIVATE_NIC_DEV(dev);
 
+	/* Queue depth must be equal to queue 0 */
+	if (qid != 0) {
+		PMD_DRV_LOG(WARNING, "rxq%u depth:%u is not equal to queue0 depth:%u.\n",
+			qid, nb_desc, nic_dev->rxqs[0]->q_depth);
+		nb_desc = nic_dev->rxqs[0]->q_depth;
+	}
 	/* Queue depth must be power of 2, otherwise will be aligned up */
 	rq_depth = (nb_desc & (nb_desc - 1)) ?
 		((u16)(1U << (ilog2(nb_desc) + 1))) : nb_desc;
@@ -852,6 +858,12 @@ static int hinic3_tx_queue_setup(struct rte_eth_dev *dev, uint16_t qid,
 	nic_dev = HINIC3_ETH_DEV_TO_PRIVATE_NIC_DEV(dev);
 	hwdev = nic_dev->hwdev;
 
+	/* Queue depth must be equal to queue 0 */
+	if (qid != 0) {
+		PMD_DRV_LOG(WARNING, "txq%u depth:%u is not equal to queue0 depth:%u.\n",
+			qid, nb_desc, nic_dev->txqs[0]->q_depth);
+		nb_desc = nic_dev->txqs[0]->q_depth;
+	}
 	/* Queue depth must be power of 2, otherwise will be aligned up */
 	sq_depth = (nb_desc & (nb_desc - 1)) ?
 		   ((u16)(1U << (ilog2(nb_desc) + 1))) : nb_desc;
-- 
2.45.1.windows.1

