From 38fbaea33486a2bf2a545a7f81ea30d4af585c20 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E6=B1=A4=E6=B5=B7=E8=95=B4?= <19855679967@163.com>
Date: Tue, 22 Jul 2025 11:58:54 +0800
Subject: [PATCH] 0005-support-tx-udp-tcp-csum-hw-offload

---
 drivers/net/hinic3/hinic3_pmd_tx.c | 90 ++++++++++++++++++++++++++++++
 1 file changed, 90 insertions(+)

diff --git a/drivers/net/hinic3/hinic3_pmd_tx.c b/drivers/net/hinic3/hinic3_pmd_tx.c
index b2199e9dea..6e784de0b5 100644
--- a/drivers/net/hinic3/hinic3_pmd_tx.c
+++ b/drivers/net/hinic3/hinic3_pmd_tx.c
@@ -323,6 +323,93 @@ int hinic3_tx_done_cleanup(void *txq, u32 free_cnt)
 	return hinic3_xmit_mbuf_cleanup(tx_queue, try_free_cnt);
 }
 
+static inline void hinic3_calculate_tcp_checksum(struct rte_mbuf *mbuf,
+					u16 inner_l3_offset)
+{
+	struct rte_ipv4_hdr *ipv4_hdr;
+	struct rte_ipv6_hdr *ipv6_hdr;
+	struct rte_tcp_hdr *tcp_hdr;
+	uint64_t ol_flags = mbuf->ol_flags;
+
+	if (ol_flags & HINIC3_PKT_TX_IPV4) {
+		ipv4_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_ipv4_hdr *,
+							inner_l3_offset);
+
+		if (ol_flags & HINIC3_PKT_TX_IP_CKSUM)
+			ipv4_hdr->hdr_checksum = 0;
+
+		tcp_hdr = (struct rte_tcp_hdr *)((char *)ipv4_hdr +
+						mbuf->l3_len);
+		tcp_hdr->cksum = rte_ipv4_phdr_cksum(ipv4_hdr, ol_flags);
+	} else {
+		ipv6_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_ipv6_hdr *,
+							inner_l3_offset);
+		tcp_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_tcp_hdr *,
+							(inner_l3_offset +
+							mbuf->l3_len));
+		tcp_hdr->cksum = rte_ipv6_phdr_cksum(ipv6_hdr, ol_flags);
+	}
+
+	return;
+}
+
+static inline void hinic3_calculate_udp_checksum(struct rte_mbuf *mbuf,
+					u16 inner_l3_offset)
+{
+	struct rte_ipv4_hdr *ipv4_hdr;
+	struct rte_ipv6_hdr *ipv6_hdr;
+	struct rte_udp_hdr *udp_hdr;
+	uint64_t ol_flags = mbuf->ol_flags;
+
+	if (ol_flags & HINIC3_PKT_TX_IPV4) {
+		ipv4_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_ipv4_hdr *,
+							inner_l3_offset);
+
+		if (ol_flags & HINIC3_PKT_TX_IP_CKSUM)
+			ipv4_hdr->hdr_checksum = 0;
+
+		udp_hdr = (struct rte_udp_hdr *)((char *)ipv4_hdr +
+						mbuf->l3_len);
+		udp_hdr->dgram_cksum = rte_ipv4_phdr_cksum(ipv4_hdr, ol_flags);
+	} else {
+		ipv6_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_ipv6_hdr *,
+							inner_l3_offset);
+		udp_hdr = rte_pktmbuf_mtod_offset(mbuf, struct rte_udp_hdr *,
+							(inner_l3_offset +
+							mbuf->l3_len));
+		udp_hdr->dgram_cksum = rte_ipv6_phdr_cksum(ipv6_hdr, ol_flags);
+	}
+
+	return;
+}
+
+static inline void hinic3_calculate_checksum(struct rte_mbuf *mbuf,
+					u16 inner_l3_offset)
+{
+	uint64_t ol_flags = mbuf->ol_flags;
+
+	switch (ol_flags & HINIC3_PKT_TX_L4_MASK) {
+		case HINIC3_PKT_TX_UDP_CKSUM:
+			hinic3_calculate_udp_checksum(mbuf, inner_l3_offset);
+			break;
+
+		case HINIC3_PKT_TX_TCP_CKSUM:
+			hinic3_calculate_tcp_checksum(mbuf, inner_l3_offset);
+			break;
+
+		case HINIC3_PKT_TX_SCTP_CKSUM:
+			/* Sctp csum no need to calculate pseudo-header */
+			break;
+
+		default:
+			if (ol_flags & HINIC3_PKT_TX_TCP_SEG)
+				hinic3_calculate_tcp_checksum(mbuf, inner_l3_offset);
+			break;
+	}
+
+	return;
+}
+
 /*lint -e40*/
 static int hinic3_tx_offload_pkt_prepare(struct rte_mbuf *mbuf,
 					 u16 *inner_l3_offset)
@@ -360,6 +447,9 @@ static int hinic3_tx_offload_pkt_prepare(struct rte_mbuf *mbuf,
 		*inner_l3_offset = mbuf->l2_len;
 	}
 
+	/* Process the pseudo-header checksum */
+	hinic3_calculate_checksum(mbuf, *inner_l3_offset);
+
 	return 0;
 }
 /*lint +e40*/
-- 
2.45.1.windows.1

